### ISSUE TITLE
[Epic] Strict serializable performance investigation

### ISSUE THREAD

**Issue by chaas** [MaterializeInc/materialize#19800]
Investigative work to answer the questions: Why are user queries blocked in serializable mode? How often are they blocked in strict serializable mode? Is strict serializability too slow and a bad user experience? Would session timelines be a better default?

We designed the strict serializable isolation level to be _correct_ without much focus on performance. If customers need queries to be performant, we recommend they switch to serializable isolation level. However, we default users into the strict serializable level, and would prefer to continue doing so because it gives users the experience they'd "expect" in a database.

The goal of this investigation is to understand:
1) the amount of latency introduced by strict serializable mode (e.g. amount of blocking b/c of timestamp logic)
2) where the latency is coming from
3) whether there are things we can do to improve the latency once we understand the source of it (while still maintaining the correctness of the strict serializable implementation)

Based on the results of the investigation above, we can either work on the strict serializable performance improvements discovered, or consider implementing session-strict serializable (session timelines) instead and using that as the default.

Recent example of a user experiencing slow performance in strict serializable, and much faster performance in serializable https://materializeinc.slack.com/archives/C03648YB34N/p1685556431488899. Feel free to reach out to field eng to learn more about their setup.

**Comment by andrewrodriguez-m** [comment]
Wanted to give an update on the work I've been doing. So far I've been investigating why explain timestamp sometimes responds with 'can respond immediately: false'. My assumption was that if we can respond immediately in strict serializable then it's equivalent to serializable: we've found a timestamp where we can query all of our sources and the rest should be the same. 


## Background / Notes from Matt and Joe
 - We get serializable (things happen in some order) for free; strict serializable is serializable + linearizable (if a tx sees timestamp ts, any tx that happens after must see at least timestamp ts). The part of the system that system guarantees linearizable is using the oracle ts for all reads. 
 - Oracle ts: used for reads, for writes only used for tables. It’s system clock with a strictly increasing bit at the end (the millisecond). Means there’s a hard cap of 1k writes/s (including DDL changes).
 - Everything has a `since` and a `upper`. You can query at times `[since, upper)`. Things before `since` have been compacted away; there's considerable freedom in delaying a compaction (we do this frequently using read holds during long running queries). `since < upper` almost always true but it's not an invariant: `since >= upper` means that we don't have a valid timestamp to query at because we've compacted away all the valid ones. 
 - Sources use system clock, this clock doesn't have to agree with the timestamp oracle for correctness because there are no consistency guarantees for sources/sinks. If the source's clock is too far into the future there would either be query lag because the timestamp chosen by the coordinator would be greater than the `since` or we'd read staleish data (not a problem because there's no guarntee on freshness). If the source's clock is too far in the past there'll be lag since the coordinator's timestamp will be greater than the source's `upper`. We use [chrony](https://chrony.tuxfamily.org/), a ntp client, to make sure we're close enough.
 - Source updates: Once a second (pretend we're at time `t`), we take all the updates that were generated in the last second (from `t-1 sec` to `t`) and give them a timestamp of `t`. Then we move our `upper` to `t+1 sec` (our granularity is 1 millisecond so effectively we can query to up `t+999ms`) and, less importantly, our `since` to `t`. The effect: we're only minting timestamps once a second, they'll be at most a second of lag between seeing new data from a source and being able to query it, we should never be waiting for a source: at switchover time `t`, the source is queryable at `[t, t+1 sec)`; this is the state of the world until `t+1` when the source becomes queryable at `[t+1 sec, t+2 sec)`. 
 - Could be interesting to get a visualization of how/when `since`s and `upper`s move. I ended up adding a bunch of `tracing::info`s to accomplish a poor man's version of this. I like the idea although one downside is that it can't answer "why didn't the `upper` move earlier?" for ex.


## Experiment

Created load generator source (`create source lg from load generator auction for all tables`) on my dev machine and then queried it locally with timing on. Seeing consistent results (30-80ms for just the select). A sample explain timestamp:
```
materialize=> begin; explain timestamp for select count(*) from bids; select mz_now(); rollback;
BEGIN
Time: 0.647 ms
                                 Timestamp                                 
---------------------------------------------------------------------------
                 query timestamp: 1686096498586 (2023-06-07 00:08:18.586) +
           oracle read timestamp: 1686096498586 (2023-06-07 00:08:18.586) +
 largest not in advance of upper: 1686096499000 (2023-06-07 00:08:19.000) +
                           upper:[1686096499001 (2023-06-07 00:08:19.001)]+
                           since:[1686096498000 (2023-06-07 00:08:18.000)]+
         can respond immediately: true                                    +
                        timeline: Some(EpochMilliseconds)                 +
                                                                          +
 source materialize.public.bids (u3, storage):                            +
                   read frontier:[1686096498000 (2023-06-07 00:08:18.000)]+
                  write frontier:[1686096499001 (2023-06-07 00:08:19.001)]+
 
(1 row)

Time: 3.425 ms
    mz_now     
---------------
 1686096498586
(1 row)

Time: 5.469 ms
ROLLBACK
Time: 0.393 ms
```

I wanted to see if I could get "can respond immediately: false" in a simple query like this. When I ran `explain timestamp for select count(*) from bids` often I'd get "can respond immediately: false" like 5% of the time. I added logs and believe that the cause is that sometimes we get a timestamp that's right above the source's `upper` and need to wait a little before we get the updated `upper`.

In the description above about how we update source's `upper`s, we assume that there's an instantaneous transition from being queryable at `[t, t+1 sec)` to `[t+1 sec, t+2 sec)` at time `t+1 sec`. This can't happen exactly and is the cause for the "can respond immediately: false" that I found when querying the `bids` table. Luckily the problem resolves itself quickly: the timestamp oracle selects some `t+1 sec+x` and we have to wait until the source updates its `upper` to include `x`. In my tests the source was already doing this so the update came pretty quickly.

An example:
```
materialize=> explain timestamp for select count(*) from bids;
                                 Timestamp                                 
---------------------------------------------------------------------------
                    current time: 1686266714264                           +
                 query timestamp: 1686266714107 (2023-06-08 23:25:14.107) +
           oracle read timestamp: 1686266714107 (2023-06-08 23:25:14.107) +
 largest not in advance of upper: 1686266714000 (2023-06-08 23:25:14.000) +
                           upper:[1686266714001 (2023-06-08 23:25:14.001)]+
                           since:[1686266713000 (2023-06-08 23:25:13.000)]+
         can respond immediately: false                                   +
                        timeline: Some(EpochMilliseconds)                 +
                                                                          +
 source materialize.public.bids (u122, storage):                          +
                   read frontier:[1686266713000 (2023-06-08 23:25:13.000)]+
                  write frontier:[1686266714001 (2023-06-08 23:25:14.001)]+
 
(1 row)
```

and added logs:
```
cluster-u14-replica-16: 2023-06-08T23:25:13.285701Z TRACE mz_storage::source::source_reader_pipeline: timely-0 remap(u126) minted new bindings: updates=[(MzOffset { offset: 39 }, 1686266714000, -1), (MzOffset { offset: 40 }, 1686266714000, 1)] source_upper={40} trace_upper={1686266714001}
cluster-u14-replica-16: 2023-06-08T23:25:13.285817Z  INFO mz_storage::storage_state: sending new frontier uppers: [(User(125), Antichain { elements: [1686266714001] })]
cluster-u14-replica-16: 2023-06-08T23:25:13.285971Z TRACE mz_storage::source::source_reader_pipeline: timely-0 reclock(u126) received remap batch: updates=[(MzOffset { offset: 39 }, 1686266714000, -1), (MzOffset { offset: 40 }, 1686266714000, 1)] upper={1686266714001}
cluster-u14-replica-16: 2023-06-08T23:25:13.286026Z TRACE mz_storage::source::source_reader_pipeline: timely-0 reclock(u126): processed 1, skipped 0 messages
cluster-u14-replica-16: 2023-06-08T23:25:13.286045Z TRACE mz_storage::source::source_reader_pipeline: timely-0 reclock(u126) downgrading timestamper: since={1686266714001}environmentd: 2023-06-08T23:25:13.299440Z  INFO mz_storage_client::controller: setting write frontier for u122 to from Antichain { elements: [1686266713001] } to Antichain { elements: [1686266714001] }
environmentd: 2023-06-08T23:25:13.299460Z  INFO mz_storage_client::controller: updates={User(122): ChangeBatch { updates: [(1686266712000, -1), (1686266713000, 1)], clean: 2 }} msg="update_read_capabilities"
cluster-u14-replica-16: 2023-06-08T23:25:13.300434Z  INFO mz_storage::source::generator: giving ((5, Ok(SourceMessage { upstream_time_millis: None, key: (), value: Row{[Int64(70), Int64(2763), Int64(7), Int32(90), TimestampTz(CheckedTimestamp { t: 2023-06-08T23:25:10.296Z })]}, headers: None })), MzOffset { offset: 40 }, 1) at cap: Capability { time: MzOffset { offset: 40 }, internal: "..." }
cluster-u14-replica-16: 2023-06-08T23:25:13.300846Z TRACE mz_storage::source::source_reader_pipeline: timely-0 remap(u126) received source upper: {41}
cluster-u14-replica-16: 2023-06-08T23:25:13.300869Z TRACE mz_storage::source::source_reader_pipeline: timely-0 reclock(u126) received source progress: source_upper={41}
...
environmentd: 2023-06-08T23:25:14.264466Z  INFO mz_adapter::coord::timestamp_selection: id_bundle=CollectionIdBundle { storage_ids: {User(122)}, compute_ids: {} } msg="determine_timestamp_for"
cluster-u14-replica-16: 2023-06-08T23:25:14.290451Z TRACE mz_storage::source::source_reader_pipeline: timely-0 remap(u126) minted new bindings: updates=[(MzOffset { offset: 40 }, 1686266715000, -1), (MzOffset { offset: 41 }, 1686266715000, 1)] source_upper={41} trace_upper={1686266715001}
cluster-u14-replica-16: 2023-06-08T23:25:14.290618Z  INFO mz_storage::storage_state: sending new frontier uppers: [(User(125), Antichain { elements: [1686266715001] })]
environmentd: 2023-06-08T23:25:14.291354Z  INFO mz_storage_client::controller: setting write frontier for u125 to from Antichain { elements: [1686266714001] } to Antichain { elements: [1686266715001] }
cluster-u14-replica-16: 2023-06-08T23:25:14.322645Z  INFO mz_storage::storage_state: sending new frontier uppers: [(User(122), Antichain { elements: [1686266715001] })]
environmentd: 2023-06-08T23:25:14.323096Z  INFO mz_storage_client::controller: id=User(122) collection=CollectionState { description: CollectionDescription { desc: RelationDesc { typ: RelationType { column_types: [ColumnType { scalar_type: Int64, nullable: false }, ColumnType { scalar_type: Int64, nullable: false }, ColumnType { scalar_type: Int64, nullable: false }, ColumnType { scalar_type: Int32, nullable: false }, ColumnType { scalar_type: TimestampTz, nullable: false }], keys: [[0]] }, names: [ColumnName("id"), ColumnName("buyer"), ColumnName("auction_id"), ColumnName("amount"), ColumnName("bid_time")] }, data_source: Other, since: None, status_collection_id: Some(System(451)) }, read_capabilities: MutableAntichain { updates: ChangeBatch { updates: [(1686266713000, 1)], clean: 1 }, frontier: [1686266713000], changes: ChangeBatch { updates: [], clean: 0 } }, implied_capability: Antichain { elements: [1686266713000] }, read_policy: Multiple([ValidFrom(Antichain { elements: [1686266714107] }), LagWriteFrontier]), storage_dependencies: [User(125)], write_frontier: Antichain { elements: [1686266714001] }, collection_metadata: CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///Users/andrewrodriguez/src/m/mzdata/persist/blob", consensus_uri: "postgres://root@localhost:26257/materialize?options=--search_path=consensus" }, remap_shard: None, data_shard: ShardId(8f7f023f-b0a0-4ebe-84d5-48e10de0512f), status_shard: Some(ShardId(64a96dee-70da-495a-985f-fa6038d0e0c7)), relation_desc: RelationDesc { typ: RelationType { column_types: [ColumnType { scalar_type: Int64, nullable: false }, ColumnType { scalar_type: Int64, nullable: false }, ColumnType { scalar_type: Int64, nullable: false }, ColumnType { scalar_type: Int32, nullable: false }, ColumnType { scalar_type: TimestampTz, nullable: false }], keys: [[0]] }, names: [ColumnName("id"), ColumnName("buyer"), ColumnName("auction_id"), ColumnName("amount"), ColumnName("bid_time")] } } } msg="update_write_frontiers"
environmentd: 2023-06-08T23:25:14.323168Z  INFO mz_storage_client::controller: setting write frontier for u122 to from Antichain { elements: [1686266714001] } to Antichain { elements: [1686266715001] }
environmentd: 2023-06-08T23:25:14.323191Z  INFO mz_storage_client::controller: updates={User(122): ChangeBatch { updates: [(1686266713000, -1), (1686266714000, 1)], clean: 2 }} msg="update_read_capabilities"
```

We can see that at time `4264` (the last four digits of the milliseconds since epoch and the number of milliseconds in the log lines) we pick a timestamp (in function `determine_timestamp_for`). This timestamp (also `4264`) is greater than the source's upper `4001` so we can't respond immediately. We wait until `4323` for the source to have data for timestamp `4264`. This isn't terrible (59ms) and doesn't explain the really terrible latency seen with strict seriazability that has been reported. This is also running on my local machine where there's no clock skew or network latency between clusterd and environmentd. Sometimes during these instances there would be updates to system tables around the same time in the logs. Although in theory I think this lag can happen with no contention because there's no overlap between the two timestamp ranges (`[t, t+1 sec)` to `[t+1 sec, t+2 sec)`), I assume any work the coordinator is doing would make this worse. 

I tried to reproduce the problem on staging but rarely got "can respond immediately: false". The average time (just using `\timing on`) was 100-200ms, some part of which is the network latency between my home and us-east-1. 

## Potential solutions, thoughts
 - Future work:
   - all the investigation so far has been querying a load generator source. We can get arbitrarily bad lag in a tower of materialized views since the height of the tower determines the amount of work mz needs to do to update the top most materialized view. Ideas: for "reasonable" sized towers of size 1-4, what does our lag look like? Future work should also include tackling a tall tower of materialized views directly.
   - quantify the busyness of the coordinator on the lag. I assumed the busier the coordinator, the more lag there would be because the coordinator might get frontier updates later than it should have. One idea is to either send some background query traffic or create some synthetic busy work for the coordinator to do.

 - Solutions: [Strict seriazability](https://jepsen.io/consistency/models/strict-serializable) says that "operations appear to have occured in some order, consistent with the real-time ordering of those operations". We have control over the real-time ordering ordering by lagging the acknowledgement of writes (we could also lag the start). This is covered in #13778 and #15706. To me this seems that this would enable us to say we have strict serializability but doesn't give the user power. Imagine the user has some web backend where they have a process that does some reads and writes to the database. They would like to avoid the weird anomalies you can get with weaker consistency models but if there are enough writes they'll be forced to do these writes in parallel which introduces its own set of potential anomalies. Sequential consistency (discussed more generally in #14052) would be a good fit here because the problem is so session focused. I'm not sure about how we can expose timelines that don't overwhelm the user while still giving them the ability to control the latency/freshness/consistency tradeoff. One avenue could be to tease apart the global consistency properties ("this user/workload should be isolated from this other user/workload") from the more local ones ("this session has x guarantee"). Not sure what to do when these conflict like when two objects are queried together but are also supposed to be isolated from each other.

**Comment by jkosh44** [comment]
Very nice write-up!

> Recent example of a user experiencing slow performance in strict serializable, and much faster performance in serializable https://materializeinc.slack.com/archives/C03648YB34N/p1685556431488899. 

It would be very useful/illuminating if we find out exactly what these queries were and the object definitions involved in the queries.

> Means there’s a hard cap of 1k writes/s (including DDL changes).

We allow DDL to bypass this restriction and write at timestamps larger than the current time. It has the potential for some negative performance impliciations though.

> If the source's clock is too far into the future there would either be query lag because the timestamp chosen by the coordinator would be greater than the since or we'd read staleish data (not a problem because there's no guarntee on freshness).

I think this is slightly off. If the source's clock is further into the future than the timestamp oracle's clock then:

- In serializable mode we may have query lag when joining this source (`A`) with another source (`B`) that doesn't have a clock in the future. The timestamp chosen would most likely be equal to the `since` of `A` and we need to wait for the `upper` of `B` to catch up to the `since` of `A`.
- In strict serializable mode (without any joins), the timestamp chosen would be `max(since, oracle ts)`. 
  - If we choose `since` then there will be query lag because we can't return a result until `oracle ts` is advanced to `since`. The results will also likely be stale.
  - If we choose `oracle ts` there won't be any query lag but the result will likely be stale.

> In the description above about how we update source's uppers, we assume that there's an instantaneous transition from being queryable at [t, t+1 sec) to [t+1 sec, t+2 sec) at time t+1 sec. This can't happen exactly and is the cause for the "can respond immediately: false" that I found when querying the bids table. Luckily the problem resolves itself quickly: the timestamp oracle selects some t+1 sec+x and we have to wait until the source updates its upper to include x. In my tests the source was already doing this so the update came pretty quickly.

I agree with this analysis. One potential mitigation is to start the update before `t` to increase the chances that the transition is complete before queries start using `t` as a timestamp.

**Comment by andrewrodriguez-m** [comment]
> > Means there’s a hard cap of 1k writes/s (including DDL changes).
> 
> We allow DDL to bypass this restriction and write at timestamps larger than the current time. It has the potential for some negative performance impliciations though.

Ah thanks, fixed inline.

> > In the description above about how we update source's uppers, we assume that there's an instantaneous transition from being queryable at [t, t+1 sec) to [t+1 sec, t+2 sec) at time t+1 sec. This can't happen exactly and is the cause for the "can respond immediately: false" that I found when querying the bids table. Luckily the problem resolves itself quickly: the timestamp oracle selects some t+1 sec+x and we have to wait until the source updates its upper to include x. In my tests the source was already doing this so the update came pretty quickly.
> 
> I agree with this analysis. One potential mitigation is to start the update before `t` to increase the chances that the transition is complete before queries start using `t` as a timestamp.

Yeah, in the current design the staleness from getting an update to the source to being able to query it is between 0 and 1 seconds. If we make it between 0.5 and 1.5 seconds we give ourselves 0.5 seconds of leeway to make the update.

**Comment by teskje** [comment]
I hope this is not randomizing, but I have been wondering how much of the latency issues observed in cloud might be caused by S3 connection issues. In the past these were really noticable because persist has configured a timeout of 60s for S3 requests and so you would sometimes have to wait that long for your results (https://github.com/MaterializeInc/materialize/issues/17227). Today the timeout is still several seconds, so definitely noticeable.

My suspicion is that with 'strict serializable' we run into these timeouts more because we sometimes have to wait for the updates for our chosen time to arrive from persist, while with 'serializable' we can usually just take what's already in the index. But that could also be totally wrong.

**Comment by jkosh44** [comment]
> My suspicion is that with 'strict serializable' we run into these timeouts more because we sometimes have to wait for the updates for our chosen time to arrive from persist, while with 'serializable' we can usually just take what's already in the index. But that could also be totally wrong.

I think this sounds totally reasonable. I would not be surprised at all if the lag in the transition from `[t, t+1 sec)` to `[t+1 sec, t+2 sec)` is exacerbated in cloud due to S3 connection issues and a general increase in network communication.

**Comment by benesch** [comment]
> Very nice write-up!

Seconding this sentiment!

I'm afraid I've not been able to spend as much time with this issue as I would have liked, and I only have a loose collection of thoughts for you as a result. I hope they are still somewhat useful.

* Querying the source directly means re-reading from S3 on every query. That's going to be fairly different from what customers are doing, which is reading out of an index. I recommend creating an indexed view on top of the source and querying that view instead: `CREATE VIEW v AS SELECT count(*) FROM src; CREATE DEFAULT INDEX ON v; SELECT * FROM v;`.
* We have a `bin/scratch` script that makes it very easy to spin up EC2 instances for development/experiments: `bin/scratch create dev-box; bin/scratch ssh <instance-id>` will get you started. That way you can query your us-east-1 staging Materialize environment with just a few milliseconds of latency.
* Let me see if I understand the glitch with source timestamping. At the start of a second (call it `t`), the source needs to commit all accumulated updates for the past second at time `t`. That might take a few hundred milliseconds, since it requires committing the data to S3. During that time, the source is (likely) unqueryable in strict serializable mode, because the timestamp oracle is likely to pick e.g. `t + 50ms`, which is not yet available. Once the data is committed, it becomes available _for the remainder of the second_, because the source's upper advances to `t + 1s`. 

  Is it right, then, that it's the S3 commit time that could be used to determine how "early" to commit data? E.g., if the p99 of writing to S3 is 300ms, we could choose to commit data for logical time `t` at wall clock time `t - 300ms`, and then 99% of the time the data would be committed by wall clock time `t`?
* How reliable of an indicator is "can respond immediately: false"? It reflects the coordinator's view of the source frontiers, but the reality in the `clusterd`s may have different frontiers. And I think that's actually all that matters? If I understand correctly, if a cluster is slow to report its frontiers back to the coordinator, "can respond immediately" might be false, even though the cluster may actually respond immediately.
* I'd be curious to see the results of an experiment where you tie the client-observed latencies for the `SELECT` query to the amount of time the query spend waiting in `clusterd` land. For slow queries, how long are they waiting in `clusterd` for the query timestamp to become available and how often is "can respond immediately" actually false? For fast queries, how long are they waiting in `clusterd` for the query timestamp to become available and how often is  "can respond immediately" actually true?

**Comment by mjibson** [comment]
For the production users who are experiencing this, are their queries referencing indexes, materialized views, or both? Are they in multi statement transactions that do timedomain stuff (determine a timestamp based on all other objects in the schema? (erp, due to #19743 we haven't done this since early May, but will again soon)).

**Comment by benesch** [comment]
> > My suspicion is that with 'strict serializable' we run into these timeouts more because we sometimes have to wait for the updates for our chosen time to arrive from persist, while with 'serializable' we can usually just take what's already in the index. But that could also be totally wrong.
> 
> I think this sounds totally reasonable. I would not be surprised at all if the lag in the transition from `[t, t+1 sec)` to `[t+1 sec, t+2 sec)` is exacerbated in cloud due to S3 connection issues and a general increase in network communication.

From discussion with @frankmcsherry today: perhaps the compaction window for sources should be a multiple of the tick interval. Say, 2s. Then there wouldn't be an repeated awkward moment where some sources have advanced but others haven't, and the coordinator can't find any time that works if you join those sources together.

That doesn't seem to be actually what's going wrong here though since there's just one source. Rather, the coordinator seems to be very interested in keeping its timestamp oracle time close to wall clock time. If it were willing to let the timestamp oracle lag behind wall clock time by, say, up to a second, that would allow it to continue to immediately serve reads for the sources that are lagging wall clock time by 200-300ms (i.e., S3 latency).

We also discussed that the phrasing "can respond immediately" has all sorts of suboptimal implications. There was some interest in rephrasing it to "coordinator introduced delay" (which is the negation of the current property).

**Comment by andrewrodriguez-m** [comment]
I wasn't able to get the raw data on what the queries were that got so much faster with serializable vs strict serializable.

One idea that's tied to some of @benesch's points is a sort of `explain timestamp` where we also run the query. Today I can run a query over and over and observe it's real performance or I can run `explain timestamp` and see if sometimes the since/uppers are weird but I can't do both. I think a nice place for this sort of data would be the upcoming query log. Then we could get a notion of "coordinator estimated wait time", "clusterd wait time for upper to meet the query timestamp", etc.

I like the idea of making the compaction window larger. The only possible downside is that in serializable mode the movement of the `since`s is what forces the readers to look at fresher data (I think?). So maybe its another freshness vs latency tradeoff.

Small experiment with stacks of materialized views: I Created a small table (`create table t(a int)`) and 1000 rows (`insert into t select * from generate_series(1, 1000)`) and ran `select count(*) from t` from us-east-1. Querying `t` directly takes ~20ms with an index and ~100-300ms without an index. Querying `t7` (7 stacked materialized views on top of `t` that just do `select * from t${n-1}` takes on average ~80-150ms normally and when there are updates to the table (including just one element added) the select query can take 400-800ms. With an index (`create default index on t7`), the normal case is 15-25ms but the unlucky select can take the same 400-800ms. Querying `t2` (2 stacked materialized views on top of `t`) with a default index produces the same normal case and the unlucky cases seems to be in the range of 150-300ms. My methodology is really poor: running psql with `\watch 0.01` and eyeballing the numbers.

**Comment by mjibson** [comment]
> One idea that's tied to some of @benesch's points is a sort of explain timestamp where we also run the query. Today I can run a query over and over and observe it's real performance or I can run explain timestamp and see if sometimes the since/uppers are weird but I can't do both.

We could add a session variable that emits a NOTICE with the EXPLAIN TIMESTAMP output when a query is executed.

**Comment by benesch** [comment]
I love that idea.

**Comment by andrewrodriguez-m** [comment]
Re making compaction window larger: thanks Joe for the clarification, I updated the text. @vmarcos: those are great results. @aalexandrov were you planning on increasing the compaction window size? I don't think I groked all your results but [strict serializable with the r/o transaction in the background results](https://github.com/MaterializeInc/letrec-bench/blob/main/analysis/notebooks/04_with_ro_transaction.ipynb) looked great

@mjibson: yeah that's a great idea

**Comment by mjibson** [comment]
Ha so we already have `emit_timestamp_notice`

```
materialize=> set emit_timestamp_notice = true;
SET

materialize=> create table t (i int);
CREATE TABLE

materialize=> select * from t;
NOTICE:  01000: query timestamp: 1689111270034
 i 
---
(0 rows)
```

This should maybe get turned into a full explain.

**Comment by andrewrodriguez-m** [comment]
Have some preliminary results on a benchmark inspired by Alexander's [letrec-bench](https://github.com/MaterializeInc/letrec-bench).

### The experimental setup
Setup an RDS instance that has account and organization fake data. We setup materialize (my staging) to consume from RDS. The blue vertical lines indicate the start of a data generation run, the red indicates the end of it. There's some time between red and blue that indicates no writes are happening. The tables I'm querying are straight postgres subsources. There are materialized views that depend on these tables but they aren't queried in this benchmark.

At the same time we're sending one query: `explain timestamp as json for select count(*) from organization, account;`. We're plotting the results we get from this explain command. We're sending queries from my machine on the west coast so we shouldn't give weight to fine latency differences. The results for running this in serialize and strict serializable are shown below:

![strict_serializable_explain](https://github.com/MaterializeInc/materialize/assets/127133603/dcebd678-58c5-4ca2-bc4c-d6e2cd9550de)

Let's start with the strict serializable results since they're a little easier to make sense of. Independent of postgres writes, our `since`s (read_frontier) and `upper`s stay in a range of [-1.5s,-.5s] and [-5s,.5s] compared to when we send the queries. Unsurprisingly we get a lot of "can respond immediately: false": that's equivalent to `upper` is less than the current time (which is clearly happening a lot). It makes sense that this graph is unrelated to data writes: we're not consistent with sources. The cool graph is to the left. We plot the same data as before but the x-axis isn't when the query was sent but rather when the query was sent mod 1sec. The first thing that jumps out is that "can response immediately: false" is much more likely to happen if you send the query between 0.2s and 0.45s. The reason is that the `upper`s are before the query send time (same as before). Interestingly at 0.3s we can either get a good `upper` or a bad one. If we "moved" the `upper`s to get minted 0.5sec earlier than they do today my guess is that there would be no "can respond immediately: false" for this benchmark. Another interesting bit: at 0.1sec the upper is behind the query send time but there are very few "can respond immediately: false". I think this is because the oracle_read_ts is less than the upper so we're good. I'm not sure why the oracle is lagging real time so much. Based on the scatter plot the oracle_read_ts seems to increment by a second once a second. One potential explanation: the only table writes are system tables that update once a second and that's what's moving the oracle_read_ts forward. I'm not confident in this explanation because the oracle_read_ts incrementing lines up very nicely with the source moving its `(since, upper)` pair so maybe it's related to the source incrementing(?).

![serializable_explain](https://github.com/MaterializeInc/materialize/assets/127133603/e3c7bd30-6b79-41dc-a385-5a15482c6844)

Next up, the serializable results. In theory the differences between serializable and strict serializable should be small: we're only doing `explain timestamp` queries. We don't get an oracle_read_ts value since it's not needed in serialiable. The graphs are the same setup as before; we're looking at the account table instead of the organization table but they both look very similar. The right graph is similar: sometimes the `upper` is less than current time which would generate "can respond immediately: false" in strict serializable. The left graph has a funny artifact where the `(since, upper)` values change their pattern. It happens while we're writing data in postgres although I'm not sure if that's pertinent. We go from `uppers` in the range of [-1s, 0s] to [0s, +1s] and back. I don't know why we don't see the same thing when running the queries in strict serializable (maybe I need to run a longer experirement?) and I don't have an explanation.

Next steps:
 - Track down how oracle_read_ts is moving and why
 - Run longer experiments to get more info on the changing since/uppers that we see in the serialiable experiment
 - The left graphs in both experiments line up with the theory that we're not giving ourselves enough time between the source updating and the controller getting the new upper. I'm going to try to mint timestamps for sources 2s into the future and set the read compaction to 2s to try and aleviate that problem.

Happy for thoughts/theories/etc!

**Comment by jkosh44** [comment]
Awesome write up!

Is the first set of graphs for strict serializable and the second set for serializable? Also is there anyway to make the colors in the legends bigger? I'm having a hard time differentiating the colors.

> The cool graph is to the left. We plot the same data as before but the x-axis isn't when the query was sent but rather when the query was sent mod 1sec.

Is it on the left or right? The graph on the right is labeled "query send time mod 1s".

> Based on the scatter plot the oracle_read_ts seems to increment by a second once a second. One potential explanation: the only table writes are system tables that update once a second and that's what's moving the oracle_read_ts forward. I'm not confident in this explanation because the oracle_read_ts incrementing lines up very nicely with the source moving its (since, upper) pair so maybe it's related to the source incrementing(?).

Two things can move the read timestamp forward:

- A write to a user or system table.
- A periodic table advancement that happens every one second.

When the timestamp is moved forward it is set to `max(system_time, current_time + 1)`. There's no coordination/communication between the timestamp oracle advancing and sources advancing.

**Comment by andrewrodriguez-m** [comment]
> Is the first set of graphs for strict serializable and the second set for serializable? Also is there anyway to make the colors in the legends bigger? I'm having a hard time differentiating the colors.
That's right. Lemme see what I can do about the legend.

> > The cool graph is to the left. We plot the same data as before but the x-axis isn't when the query was sent but rather when the query was sent mod 1sec.
> 
> Is it on the left or right? The graph on the right is labeled "query send time mod 1s".
Doh, you're correct! Fixed in the text

> > Based on the scatter plot the oracle_read_ts seems to increment by a second once a second. One potential explanation: the only table writes are system tables that update once a second and that's what's moving the oracle_read_ts forward. I'm not confident in this explanation because the oracle_read_ts incrementing lines up very nicely with the source moving its (since, upper) pair so maybe it's related to the source incrementing(?).
> 
> Two things can move the read timestamp forward:
> 
> * A write to a user or system table.
> * A periodic table advancement that happens every one second.
> 
> When the timestamp is moved forward it is set to `max(system_time, current_time + 1)`. There's no coordination/communication between the timestamp oracle advancing and sources advancing.
I see, that makes sense. When I run `copy (subscribe (select * from mz_internal.mz_compute_frontiers_per_worker)) to stdout;` I see that mz_compute_frontiers is pushing new updates every 100ms. I assume that would mean that we'd increment the oracle every 100ms?

**Comment by jkosh44** [comment]
> When I run `copy (subscribe (select * from mz_internal.mz_compute_frontiers_per_worker)) to stdout;` I see that mz_compute_frontiers is pushing new updates every 100ms. I assume that would mean that we'd increment the oracle every 100ms?

`mz_compute_frontiers` is an introspection source (sometimes referred to as a Log in the code for legacy reasons). Introspection sources are treated the same as normal sources in that they don't use or affect the timestamp oracle. Storage collections (or whatever they're called these days) are treated the same. Only system objects that are referred to as `BuiltinTable` in `builtin.rs` will push the timestamp oracle forward when they're written to.

**Comment by mjibson** [comment]
These graphs and summary are great.

> The right graphs in both experiments line up with the theory that we're not giving ourselves enough time between the source updating and the controller getting the new upper. I'm going to try to mint timestamps for sources 2s into the future and set the read compaction to 2s to try and aleviate that problem.

Assuming those changes do what you hope, I'd be interested in seeing similar graphs but with stacks of dependent MVs so we can also observe that lag, and the effect it may have on different kinds of queries.

**Comment by aalexandrov** [comment]
> Assuming those changes do what you hope, I'd be interested in seeing similar graphs but with stacks of dependent MVs so we can also observe that lag, and the effect it may have on different kinds of queries.

@andrewrodriguez-m please ping me if you want some help when changing the workload to accommodate for this requirement.

**Comment by andrewrodriguez-m** [comment]
With [custom code](https://github.com/MaterializeInc/materialize/pull/20835/files/2ba63e0239cf547fa3fbebe7970214e8d2908397) that sets the read hold to 5 secs and mints write timestamps one second ahead of normal we get the expected graphs. 

Strict serializable:
![strict_serializable_5s_read_hold_1s_write_ahead](https://github.com/MaterializeInc/materialize/assets/127133603/d97632db-038b-4ddf-8941-bf94f036c8d6)

Serializable:
![serializable_5s_read_hold_1s_write_ahead](https://github.com/MaterializeInc/materialize/assets/127133603/41e4a6ac-bc8c-4eba-9fd5-704f658211be)

We see the same strange behavior where since/uppers move around in strict serializable. According to @guswynn on slack: "we bump the source ts every second (using mz_now() basically), and then process the data (for postgres, this is not much, just moving it around), then write it out to persist (in batches) at that ts. I suspect that difference here is the processing delay". I'm still confused why this only seems to happen when the queries are sent in a session with strict serialiazable but it seems less pressing to explore.

In strict serializable we still see one "can respond immediately: false" although it looks like an anomaly.

Tradeoffs:
 - Increasing the read hold seems pretty innocuous. Up to 5 seconds of updates can accumulate w/o being able to consolidate. I don't think it's a big deal
 - Minting writes in the future for sources. The freshest data you can query is now at least 1-2 seconds (just a lower bound, we have no control over the lag between real source and mz) vs 0-1 seconds. This doesn't seem so bad although it does eat into the value prop of mz.
 - Minting writes in the future for tables. This is trickier than doing it for sources because it would move the oracle timestamp forward. The reason minting timestamps a second in advance works is that the oracle ts is close to wall clock time and the source's timestamp is always [1, 2) seconds greater. A table write would move the oracle ts such that the source's timestamp is [0, 1) seconds greater, like today, and we'd have the same issues.

Different minting timestamp behavior for tables vs sources seems weird. One way to square them is to imagine that tables are backed by a postgres source. When we do a write to a table, we write into the postgres source and wait until the wall clock time is at least as big as its assigned timestamp before returning the user. (Generally, we'd wait 2 seconds). This is the same idea that was proposed in #15706 and #13778.

Next steps:
 - I'll try an experiment with some stacks of materialized views on top of the postgres sources. See how many we need to get more "can respond immediately: false" keeping the minting timestamps 1 sec in the future.

**Comment by andrewrodriguez-m** [comment]
I was running a materialize view experiment and was noticing the since/uppers on materialized views that used the `organizations` table were the same as the `organization` tables. That table is populated once at the beginning of the run... we were looking at the since/uppers of a table that didn't get updates. I reran the previous post's experiment with the `transaction` table which does get updates the whole benchmark run.

Serializable:
![txn_s](https://github.com/MaterializeInc/materialize/assets/127133603/4d2364fc-63b7-40ed-b183-99a4b4b604ff)

Strict serializable:
![txn_ss](https://github.com/MaterializeInc/materialize/assets/127133603/898157ef-b718-4673-8fbb-70d9afe50f27)

(I assume the outlier above is due to the fact that we were sending queries for a chunk of time. The current experimental setup is vulnerable to coordinated omission -- we only send queries when there is a worker in the pool available. For now I'm going to ignore it because I'm more interested in the general trend of since/uppers.)

These results are in line with what we saw earlier: since/uppers assigned shouldn't be affected by whether or not we're writing to sources

**Comment by andrewrodriguez-m** [comment]
With 2 materialized views the since/uppers drift down. The setup is similar to before but we've added two materialized views: 
```
create materialized view transaction_mv1 in cluster "srv_cluster" as select * from transaction where src_id != 999999999999;
create materialized view transaction_mv2 in cluster "srv_cluster" as select * from transaction_mv1 where src_id != 999999999998;
```
. Q6 does `explain timestamp for select count(*) from transaction_mv1` and Q7 does the equivalent for `transaction_mv2`. Q5 is the same query we were running before.

Serializable:
![2mvs_txn_s](https://github.com/MaterializeInc/materialize/assets/127133603/960c9bf0-0bba-4437-9fb3-2b3c716696a4)

Strict serializable:
![2mvs_txn_ss](https://github.com/MaterializeInc/materialize/assets/127133603/bcfacf8b-736c-4fbf-839a-1fe756acd14f)

It's a bit hard to tell but the points representing q6 (1 mat view) and q7 (two mat views) have lower since/uppers; click the image to be able to zoom in. There are a few "can response immediately: false".

To try to make the difference more obvious I ran another condition where I'm measuring 7 stacked materialized views:

Serializable:
![txn_s_7mv](https://github.com/MaterializeInc/materialize/assets/127133603/36e80f9c-a548-49f5-804b-89bcea3233a2)

Strict serialiazable:
![txn_ss_7mv](https://github.com/MaterializeInc/materialize/assets/127133603/df6156dd-f54b-4ea4-991a-c1fbdeac9b4b)

Ignore the description of Q7, it's actually querying `transaction_mv7`. There are more "can respond immediately: false" although the penalty per materialized view is not 1s, it seems that with 7 we're running about 1s behind total.

**Comment by andrewrodriguez-m** [comment]
More comments on the graphs: in the first half a minute there's no difference between the 7th materialized view and the `transaction` table. I think that's because at the beginning of the benchmark we're writing `account` and `organization` data so there are no writes. For the same reason there's only a difference between the blue and red vertical lines which indicate that we're actively writing.

**Comment by mjibson** [comment]
Some questions I came up with:

1. why is can respond immediately false more between .2s and .45s? why doesn't happen between .0s and .2s?
2. oracle read ts updates at least once per second. does it happen at some random subsecond, or do we try to get it exactly at the 1.000s boundary?
3. would it matter if we changed to do the opposite of the above?
4. what is a workload that would produce can respond immediately false for serializable (which we observe happens in production)?
5. what if there were a thing doing inserts in a loop (which I believe advances oracle read ts)? would that advance oracle read frequently enough that the 0.0s - 0.2s space would decrease (or some other effect)?
6. could we add a .5s offset to source timestamp closing? from 0.0s to 0.5s, incoming source data gets assigned 1.0s. at 0.5s, 1.0s is closed. from 0.5s to 1.5s, incoming source data gets assigned 2.0s. at 1.5s, 2.0s is closed. etc. this may generally provide that source uppers are always > oracle read ts, wouldn't increase the rate at which we close timestamps, and would only increase the max date lateness to 1.5s instead of 2s. this delay should perhaps be configurable (and, one could imagine, dynamically and automatically tunable by sources themselves based on feedback from envd). one way to imagine this is that today (well, in the past) we imagined that the time it took between a source closing a timestamp and envd observing that was 0 (this was probably true in the pre-cloud days). but now it takes some hundreds of ms for a close timestamp event from storage to propagate to envd. is it reasonable to have envd send back messages to sources of the form "hey i got your close timestamp t message at time X". the source can then diff those (and get ~300ms), then adjust its next batch to initiate the close timestamp event 300ms before it otherwise would have.

**Comment by mjibson** [comment]
One other idea we discussed: Close timestamps each .5 seconds instead of each second, but still 1s ahead of wallclock. At 0.5s we close for 0.5-1.0s. At 1.0s close for 1.0-1.5s, etc. This would not worsen data freshness, but would double the amount of timely close events, which may or may not have adverse CPU affects.

**Comment by andrewrodriguez-m** [comment]
Re question 2, how does oracle ts work:

`advance_timelines_interval` ticks once a second and simulates receiving a Message::GroupCommitInitiate message. Everytime we get this message we mint a new timestamp in `group_commit_initiate` which is then applied in `group_commit_apply`. `advance_timelines_interval`'s missed tick behavior is [Burst](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Burst) so if we're behind we'll execute as many ticks as possible to catch up but then we'll try to get on the existing period. We also call `advance_timelines_interval.reset()` in `send_builtin_table_updates` and `trigger_group_commit` which resets the ticket period starting when reset is called. In both cases we do this after starting a `group_commit_initiate` invocation although in the `send_builtin_table_updates` we do the reset after the `group_commit_initiate` has completed (which may or may not have waiting on `group_commit_apply` to have completed, depending on if there were any DDL writes) whereas in `trigger_group_commit` we do the reset after writing the GroupInitiateCommit message into the internal message queue. `sent_builtin_table_updates` is called every time we do a catalog transaction and add/remove a subscribe so these are not infrequent and will change the part of the second when we update oracle ts. The resets should not delay the oracle ts from getting updated too much since it's right around the time we minted a new timestamp.

Oracle timestamp is always less than the wall clock time (once a second it's set to wall clock, the time after it's stale). Same with sources: in RemapClock we're minting the next upper as wall clock rounded down to the previous whole second. It seems interesting to align the two, if the oracle ts and the sources were consistently half a second off that's lag that we're not budgeting for.

**Comment by jkosh44** [comment]
> 4. what is a workload that would produce can respond immediately false for serializable (which we observe happens in production)?

I haven't tested this locally, but something like the following should work:


```SQL
SET transaction_isolation TO serializable;
CREATE TABLE t (a INT);
CREATE MATERIALIZED VIEW mv AS (expensive query or a stack of materialized views);


SELECT * FROM t, mv;
```

**Comment by andrewrodriguez-m** [comment]
Current proposal: https://www.notion.so/materialize/Latency-vs-Freshness-in-strict-serializable-0e7d0a2e3a424237842eb8d03dc312f3?pvs=4
